docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,798 @@
+
+If your use-case doesn't need so many levels to identify an entity,
+the unnecessary levels can be skipped with the `_` (underscore) character as follows:
+
+```
+te/<entity-id>/_/_/_/...
+```
+
+or 
+
+```
+te/_/_/_/<entity-id>/...
+```
+
+Even for the data type levels, a user is free to define those as they wish.
+But thin-edge has some pre-defined `<data-type>` subtopics for well-known types like:
+`measurements`, `alarms`, `events` and `commands`, on which it enforces some constraints.
+
+**Subtopics for predefined telemetry data**
+
+|Type|Example|
+|------|-------|
+|Measurements|`<entity_id_prefix>/mes/[/<measurement-type>]`|
+|Events|`<entity_id_prefix>/evn/<event-type>`|
+|Alarms|`<entity_id_prefix>/alr/<alarm-type>`|
+
+**Subtopics for predefined commands**
+
+|Type|Example|
+|------|-------|
+|Software List|`<entity_id_prefix>/cmd/software_list`|
+|Software Update|`<entity_id_prefix>/cmd/software_update`|
+|Configuration Snapshot|`<entity_id_prefix>/cmd/config_snapshot`|
+|Configuration Update|`<entity_id_prefix>/cmd/config_update`|
+|Firmware Update|`<entity_id_prefix>/cmd/firmware_update`|
+|Restart[^1]|`<entity_id_prefix>/cmd/restart`|
+
+[^1]: Restart would mean a device restart or a service restart based on the target entity
+
+### Entity registration
+
+Since thin-edge doesn't enforce what each device identification level means,
+an explicit registration is required to register every entity that is going to send data or receive commands.
+For example, before a measurement can be sent from a service named `collectd` from the device `Rpi1001`,
+the entity named `Rpi1001` must be registered as a `device`, and `collectd` must be registered as a `service`.
+
+An entity can be registered with thin-edge by publishing a retained message to the entity identification topic prefix
+with the entity type and other metadata that defines that entity.
+To model the example mentioned above, an entity identification topic scheme like the following can be used:
+
+```
+te/<device-namespace>/<device-id>/<service-namespace>/<service-id>
+```
+
+The `Rpi1001` device must be registered as a `device` by publishing the following retained message:
+
+```sh te2mqtt
+tedge mqtt pub -r tedge/_/Rpi1001 '{
+  "type": "device"
+}'
+```
+
+The device is registered under the default namespace indicated by `_` at the second level.
+The registration message supports additional fields like `name`, `parent`, `device-type` etc as well,
+which will be explained in detail later.
+
+Once the device is registered, the `collectd` service can be registered as a `service` type as follows:
+
+```sh te2mqtt
+tedge mqtt pub -r tedge/_/Rpi1001/_/collectd '{
+  "type": "service"
+}'
+```
+
+Once these entities are registered, they can start sending data or receive commands.
+
+Examples:
+* Device measurement: `tedge/_/Rpi1001/_/_/mes/battery_reading`
+* Service measurement: `tedge/_/Rpi1001/_/collectd/mes/cpu_usage`
```

Comment:
To publish the `telemetry` data for Rapi1001, the publisher must know the whole topic structure, right? I feel this is too long a topic.

Reply:
I'm hoping that once we establish that there are a fixed number of topic levels for everything, it would be far easier for a user to follow rather than having variable length topics for main device vs child device services.